src/environ/_environ_config.py

```diff
@@ -174,11 +174,72 @@ class Sub:
 
     You can nest your configuration as deeply as you wish.
     """
+    default = None if optional else RAISE
     return attr.ib(
-        default=None, metadata={CNF_KEY: _ConfigEntry(None, None, cls, True)}
+        default=default,
+        metadata={CNF_KEY: _ConfigEntry(None, default, cls, True)},
     )
 
 
+def __default_getter(environ, metadata, prefix, name):
+    ce = metadata[CNF_KEY]
+    if ce.name is not None:
+        var = ce.name
+    else:
+        var = "_".join(prefix + (name,)).upper()
+    log.debug("looking for env var '%s'.", var)
+    try:
+        return environ[var]
+    except KeyError:
+        raise MissingEnvValueError(var)
+
+
+def __to_config_recurse(config_cls, environ, prefixes, default=RAISE):
+    # We keep track of values we actually got from the getter vs those we set
+    # from the `ConfigEntry` default value
+    got = dict()
+    defaulted = dict()
+    missing_vars = set()
+
+    for attr_obj in attr.fields(config_cls):
+        try:
+            ce = attr_obj.metadata[CNF_KEY]
+        except KeyError:
+            continue
+        name = attr_obj.name
+
+        if ce.sub_cls is not None:
+            got[name] = __to_config_recurse(
+                ce.sub_cls, environ, prefixes + (name,), default=ce.default
```

Comment:
...but I'm pretty sure I actually want the `(name,)` bit here to be `ce.name or ce.sub_cls._prefix` instead so that the prefix of config classes used as groups can be maintained. Does that sound right?

Reply:
This comment should have read `ce.sub_cls._prefix or name` but it's also a bit more complicated than that. I've pushed a new change which captures the edge cases, let me know if anything in there looks funky.